oak_functions/lookup/README.md

```diff
@@ -0,0 +1,49 @@
+# Lookup Extension
+
+With the lookup extension, a request can do private key/value lookups in
+`Oak Functions`.
+
+The key idea is that `Oak Functions` initially loads the entire lookup data into
+the TEE and requests can then do a key/value lookup on this lookup data.
+
+We want to maintain the following invariants on the key/value lookups.
```

Comment:
I just remembered another invariant that is very important: The timing of refreshing the lookup data must not depend on user requests. This is to ensure that refreshing of lookup does not leak information.

Reply:
Yes, I mean updating the lookup data. The idea is that if the Wasm logic is able to in any way trigger updating the lookup data it can use this to signal information, since updating lookup data is externally visible.

As an example, let's say that the encrypted user request contains information about membership in some sensitive group (e.g. people suffering from some health condition). If the Wasm logic sees that the current user is a member of such a group it could signal this to the outside world by triggering a lookup data update.

Reply:
Thank you, @conradgrobler, I see. So the Wasm logic contains something `if sensistive_group.is_member("person") { update_lookup_data() }` and the Untrusted Launcher observes that the request triggered update_lookup_data. 

I think we trivially have this at the moment by providing no handle to the Wasm module to updata lookup data. 

I'll add this as an invariant, I agree we have to make sure that this is maintained. Especially as "refreshing data" is a very natural thing for someone writing Wasm modules to want.

Reply:
I don't think the Wasm logic can update the lookup data, can it?

Reply:
Or is this what we are trying to clarify here? Though it seems oddly specific to me; perhaps the invariant should be that nothing observable by the untrusted host may depend on secret data; refreshing lookup data is one of those observable things, but so are many other.

Reply:
No, it can't right now. It should never, though.

Reply:
> Though it seems oddly specific to me; perhaps the invariant should be that nothing observable by the untrusted host may depend on secret data; refreshing lookup data is one of those observable things, but so are many other.

I think it is important to enumerate the observable things we are defending against. We are not currently defending against a malicious hypervisor that monitors page accesses, so some information about the lookup information will leak. So we cannot at this point claim we are maintaining the more general invariant.

Reply:
It just seems to me that this is not an invariant about lookup data, it is an invariant about the observability of secret-dependent behaviour. But I guess it does not harm to have it here as well.